flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/SinkWriterOperator.java

```diff
@@ -195,12 +216,18 @@ public void processWatermark(Watermark mark) throws Exception {
 
     @Override
     public void endInput() throws Exception {
-        endOfInput = true;
-        sinkWriter.flush(true);
-        emitCommittables(Long.MAX_VALUE);
+        if (!this.endOfInput) {
+            this.endOfInput = true;
+            // endOfInputState is union state, so it's enough if one task adds something to it
```

Comment:
What happens to the sinkWriter when two subtasks with different endOfInputState got merged on down-scaling?  Shouldn't sink writer be be flushed? (I see it also uses operator list state ("writer_raw_states")) IIUC, it wouldn't

Reply:
You are correct that this was not correct, especially also on partially completed tasks.

I changed it to regular state and let the committer handle duplicates now. The main idea is to avoid extra transactions as much as possible in the writer and if it's not possible, let the committer clean them up anyhow.